src/components/scripts/oauth.mjs

```diff
@@ -0,0 +1,221 @@
+
+    this.#socketServer = new TCPServerSocket("::");
+  }
+
+  getOAuthAccessCode(client_id, client_secret, scopes) {
+    this.#clientId = client_id;
+    this.#clientSecret = client_secret;
+    this.#oauthScopes = scopes;
+
+    const { promise, resolve, reject } = Promise.withResolvers();
+    this.#promiseResolver = resolve;
+    this.#promiseRejecter = reject;
+
+    this.#socketServer.opened.then(
+      (server) => {
+        this.#redirectUri = `http://localhost:${server.localPort}`;
+        this.#runHttpServer();
+        this.#authenticate(this.#authentication_endpoint);
+      },
+    );
+
+    return promise;
+  }
+
+  get authentication_token() {
+    return this.#authenticationCode;
+  }
+
+  get access_token() {
+    return this.#accessCode;
+  }
+
+  get refresh_token() {
+    return this.#refreshCode;
+  }
+
+  #authenticate(endpoint) {
+    // Create <form> element to submit parameters to OAuth 2.0 endpoint.
+    let form = document.createElement("form");
+    form.setAttribute("method", "GET"); // Send as a GET request.
+    form.setAttribute("action", endpoint);
+    form.setAttribute("target", "_blank");
+
+    // Parameters to pass to OAuth 2.0 endpoint.
+    let params = {
+      client_id: this.#clientId,
+      redirect_uri: this.#redirectUri,
+      response_type: "code",
+      scope: this.#oauthScopes.join(" "),
+      include_granted_scopes: "true",
+      state: "pass-through value",
+    };
+
+    // Add form parameters as hidden input values.
+    for (let p in params) {
+      let input = document.createElement("input");
+      input.setAttribute("type", "hidden");
+      input.setAttribute("name", p);
+      input.setAttribute("value", params[p]);
+      form.appendChild(input);
+    }
+
+    // Add form to page and submit it to open the OAuth 2.0 endpoint.
+    document.body.appendChild(form);
+    form.submit();
+  }
+
+  async #runHttpServer() {
+    const server =
+      await this.#socketServer.opened;
+    const connections = server.readable.getReader();
+
+    while (true) {
+      const { value: connection, done } = await connections.read();
+
+      // Send the connection to the callback
+      if (connection) {
+        this.#connectionReceived(connection);
+      }
+
+      // Release the connection if we're done
+      if (done) {
+        connections.releaseLock();
+        break;
+      }
+    }
+
+    // Wait for the server to be closed
+    await this.#socketServer.closed;
+    this.#socketServer = undefined;
+  }
+
+  async #connectionReceived(connection) {
+    return connection.opened.then(async (socket) => {
+      const reader = socket.readable.getReader();
+      const value = await this.#readStream(reader);
+      const { success, html } = this.#generateResponse(value);
+      await this.#writeStream(socket, html);
+
+      connection.close();
+
+      if (success) {
+        const accessCode = await this.#getAccessCode();
+        this.#promiseResolver(accessCode);
+      }
+    });
```

Comment:
```suggestion
    const socket = await connection.opened;
    const reader = socket.readable.getReader();
    const value = await this.#readStream(reader);
    const { success, html } = this.#generateResponse(value);
    await this.#writeStream(socket, html);

    connection.close();

    if (success) {
      const accessCode = await this.#getAccessCode();
      this.#promiseResolver(accessCode);
    }
```

Instead of returning the promise and changing a `.then`, we can await connection.opened directly for socket and continue from there